### ISSUE TITLE
Numpy 2.0 compatibility

### ISSUE THREAD

**Issue by MatthewFlamm** [pyvista/pyvista#5533]
### Describe what maintenance you would like added.

numpy 2.0 testing is currently be tracked in #5450.  Several fixes have already been merged. This issue is for tracking remaining issues (and add any that pop up).  Details of each in further comments.

- [ ] garbage collection errors
- [x] meshio ~incompatibility and~ errors
- [ ] array casting errors (as of release candidate 2.0.0rc2)

**Comment by MatthewFlamm** [comment]
garbage collection errors (~150 tests failing)

Example error from `tests/plotting/test_actor.py::test_actor_copy_deep`.  It may be a useful comparison that `tests/plotting/test_actor.py::test_actor_from_plotter` does not have this error.

```txt
>       assert len(after) == 0, msg
E       AssertionError: Not all objects GCed:
E         vtkWeakReference: ["dict: d['dataset']", "dict: d['dataset']"]
E         vtkWeakReference: ["dict: d['dataset']", "dict: d['dataset']"]
E         vtkFloatArray: ["dict: d['VTKObject']", "dict: d['VTKObject']", '<managedbuffer object at 0x7fe309925d40>', '<managedbuffer object at 0x7fe309924740>', "dict: d['VTKObject']", "dict: d['VTKObject']"]
E         
E       assert 3 == 0
E        +  where 3 = len([<vtkmodules.vtkCommonCore.vtkWeakReference(0x55a25c494080) at 0x7fe309978ac0>, <vtkmodules.vtkCommonCore.vtkWeakReference(0x55a25bfc2030) at 0x7fe309978ee0>, <vtkmodules.vtkCommonCore.vtkFloatArray(0x55a25d0d9b10) at 0x7fe309978a60>])
```

**Comment by MatthewFlamm** [comment]
meshio.  The biggest issue here is the recent lack of maintenance in `meshio`. Without this changing, we need to consider either fully dropping `meshio` as a dependency or firewalling it to be an optional dependency (and only used for numpy <2).

https://github.com/nschloe/meshio/pull/1451 fixes import problems.

We have other errors related to `pyvista.save_meshio`, snippet below.  It is unclear whether this is an internal or meshio issue at this time.  An example is from `tests/test_meshio.py::test_meshio` and the error occurs inside the `pyvista.save_meshio` code.

```txt
            for k in range(len(data)):
>               data[k] = np.asarray(data[k])
E               TypeError: 'tuple' object does not support item assignment
```

**Comment by keurfonluu** [comment]
> meshio. The biggest issue here is the recent lack of maintenance in `meshio`. Without this changing, we need to consider either fully dropping `meshio` as a dependency or firewalling it to be an optional dependency (and only used for numpy <2).
> 
> [nschloe/meshio#1451](https://github.com/nschloe/meshio/pull/1451) fixes import problems.
> 
> We have other errors related to `pyvista.save_meshio`, snippet below. It is unclear whether this is an internal or meshio issue at this time. An example is from `tests/test_meshio.py::test_meshio` and the error occurs inside the `pyvista.save_meshio` code.
> 
> ```
>             for k in range(len(data)):
> >               data[k] = np.asarray(data[k])
> E               TypeError: 'tuple' object does not support item assignment
> ```

That concerns me as well, especially since `meshio` is used in many other Python packages. The first thing we can do is to ping @nschloe about any ETA for the next release and/or to allow other devs to maintain `meshio` in their stead (I volunteer as tribute). The alternative is to fork a new project (e.g., `meshio2`, `pvmeshio`...) that we can maintain.

**Comment by tkoyama010** [comment]
> The alternative is to fork a new project (e.g., meshio2, pvmeshio...) that we can maintain.

Or merge as module `meshio` to pyvista project.

**Comment by nschloe** [comment]
Hi guys, meshio author here. Thanks for pinging me! I'll merge Matthew's PR to make meshio build against the upcoming numpy again. That should hopefully fix the issue here.

Perhaps later there'll be time to remove the meshio dependency. That'd allow adding pyvista as dependecy to meshio, as I had thought about for a while now.

**Comment by nschloe** [comment]
[meshio 5.3.5](https://pypi.org/project/meshio/5.3.5/) released.

**Comment by tkoyama010** [comment]
Bumped in #5548

**Comment by MatthewFlamm** [comment]
It looks like the remaining errors related to `save_meshio` are likely internal.  Specifically `numpy.split` now returns a tuple rather then a list, despite not being mentioned in the numpy release/migration documentation.

@keurfonluu you implemented this section in that code
https://github.com/pyvista/pyvista/blame/33fdf89997dd748313efd2accd24b84359022fef/pyvista/core/utilities/fileio.py#L599C4-L603C6

I don't understand the usage of `numpy.split` here.  Would it be better to rewrite this or coerce to a list?

**Comment by keurfonluu** [comment]
> I don't understand the usage of `numpy.split` here. Would it be better to rewrite this or coerce to a list?

Cells in `meshio` are split by cell block, each cell block having the same cell type. Data arrays are split in the same fashion.
I believe that the fact that the new version of `numpy` returns a tuple instead of a list was indeed the issue, so coercing `numpy.split`'s output to a list should do the trick.

**Comment by MatthewFlamm** [comment]
A simple MRE for at least some of the garbage collection errors.  Note that some of this must be run somewhere inside `tests/plotting/` as the garbage collection fixture is there.  At least some of these errors don't even need a plotter involved.

```py
def test_plane():  # garbage collection error
    pv.Plane()

def test_sphere():  # no garbage collection error
    pv.Sphere()
```


If these lines are commented out, then it passes.

https://github.com/pyvista/pyvista/blob/499c39e8b20e14a83de7cd4f751bbe9132819b86/pyvista/core/utilities/geometric_objects.py#L976-L977

Also passes if they are changed to:

```py
    surf.points[:, 0] = surf.points[:, 0] * i_size
    surf.points[:, 1] = surf.points[:, 1] * j_size
```


I'm not sure whether this is some error in our pyvista_ndarray implementation.

**Comment by adeak** [comment]
The (recently appeared) array casting errors are mostly easy to handle, I've fixed almost all locally.

The garbage collection issue, on the other hand, is very very hairy. I'm still investigating.

Almost minimal memory leak example:
```python
mport gc
import pyvista as pv

def leaky():
    points = pv.Sphere().points
    points[:, 0] *= 1

for i in range(300):
    if i % 10 == 0:
        gc.collect()
        print(len(gc.get_objects()))
    leaky()
```
This keeps creating temporary arrays and checks the total number of objects tracked by the garbage collector. On NumPy 1.26.4 we see a constant value, but with 2.0.0rc2 we see a linear increase like this:
```
124575
124669
124759
124849
124939
125029
125119
125209
125299
125389
125479
125569
125659
125749
125839
125929
126019
126109
126199
126289
126379
126469
126559
126649
126739
126829
126919
127009
127099
127189
```
We gain 90 new references per 10 iterations (unless I miscalculated; the fact that this constantly increases is enough to have an issue). As you noted, this is strongly related to `__imul__()`: changing `leaky()` to contain regular multiplication makes the issue go away.

---

So here's an even more minimal example that we have a problem (on NumPy 2.0.0rc2):
```python
import gc
import weakref
import pyvista as pv

# regular mul returns an ephemeral array
wref = weakref.ref(pv.Sphere().points.__mul__(1))
gc.collect()
print(wref)
# <weakref at 0x7f7f0b5c6bb0; dead>

# imul sticks around
wref = weakref.ref(pv.Sphere().points.__imul__(1))
gc.collect()
print(wref)
# <weakref at 0x7f7f3488a4d0; to 'pyvista_ndarray' at 0x7f7f3478cdd0>
```
Again on NumPy 1.26.4 both weak references end up dead.

---

Here's the part that confuses the hell out of me currently: let's check what keeps that array alive.
```python
import weakref
import gc
import pyvista as pv

wref = weakref.ref(pv.Sphere().points.__imul__(1))
refs = gc.get_referrers(wref())
print(len(refs))  # 1
ref = refs[0]
print(ref)  # bound method pyvista_ndarray.__array_wrap__ of pyvista_ndarray(...)

print(ref.__self__ is wref())  # True
# in other words
print(wref().__array_wrap__.__self__ is wref())  # True
# but also
print(wref().__array_finalize__.__self__ is wref())  # True
print(wref().__setitem__.__self__ is wref())  # True
```

So here's the thing: it looks like we have `pyvista_ndarray` instances sticking around after calls to their `__imul__()` (which is inherited from `np.ndarray`). They stick around because they are referenced by their own `__array_wrap__()` methods. What I don't get:
1. Why doesn't the garbage collector handle an object that's only referred to by one of its methods?
2. How does `points.__array_wrap__()` hold a reference to `points`? If it were just via `__self__` that all bound methods have, we'd see more than 1 referrer, we'd have to see every method (see the last prints in the last previous code block).
3. OK, assume `points` (or, more likely, one of its clones) is still alive because `points.__array_wrap__()` references it. But then why do I see that `len(gc.get_referrers(wref().__array_wrap__))`? Perhaps it's because, as [`gc` docs says](https://docs.python.org/3/library/gc.html), "_This function will only locate those containers which support garbage collection; extension types which do refer to other objects but do not support garbage collection will not be found_", but `points` is implemented in Python...

So I'm currently getting more confused the more I look at this. I suspect the issue really has something to do with `__array_wrap__()`, but I currently have no idea what. I might end up opening a NumPy issue just in case someone can offer some hints.

**Comment by adeak** [comment]
So [we might be innocent](https://github.com/numpy/numpy/issues/26545), or some extra steps are needed now to subclass `ndarray` properly.

**Comment by adeak** [comment]
NumPy bug confirmed and fix PR already open upstream. We might want to use nightly wheels once it's fixed to be able to use CI.

**Comment by adeak** [comment]
Actually, no, we'd need nightly builds of the 2.0.0 release branch...

**Comment by MatthewFlamm** [comment]
> Actually, no, we'd need nightly builds of the 2.0.0 release branch... `main` is 2.1.0 dev now.

2.0.0 nightly wheels are not available at  https://anaconda.org/scientific-python-nightly-wheels/numpy and I cannot easily find any other sources.

**Comment by adeak** [comment]
Having given it some thought, I think the regular nightlies will do. It's true that we won't exactly be testing the 2.0.0 release that will ship soon, but the kind of issues we expect will be fixed on the main branch and backported to 2.0.0, so I think we can work with that. I'll wait to see the next nightly build to check that our gc errors are gone, then open a PR and update the numpy pre-release case to nightly.

**Comment by MatthewFlamm** [comment]
I agree with this proposal.  We will support 2.1 in the future and this allows us to get it into the CI faster, which will prevent more problems.
